data/processors/sitemap.py

```diff
@@ -0,0 +1,142 @@
+import json
+import urllib.request
+import xml.etree.ElementTree as ET
+import defusedxml.ElementTree as defusedET  # supports only parse()
+from datetime import datetime
+
+
+def generate_sitemap():
+    """ Generate a sitemap that diffs changes since to the currently online data """
+    # Load exported data. This function is intentionally not using the data object
+    # directly, but re-parsing the output file instead, because the export not
+    # export all fields. This way we're also guaranteed to have the same types
+    # (and not e.g. numpy floats).
+    with open("output/api_data.json") as f:
+        new_data = json.load(f)
+
+    # Currently online data
+    req = urllib.request.Request("https://nav.tum.sexy/cdn/api_data.json")
+    with urllib.request.urlopen(req) as resp:
+        old_data = json.loads(resp.read().decode("utf-8"))
+
+    # Each sitemap has a limit of 50MB uncompressed or 50000 entries
+    # (that means 1KB per site). We have currently about 33000 entries,
+    # so it's unlikely that we'll hit this limit without adding a lot of
+    # data. But for the case that a new type of entry is introduced, the
+    # sitemap is split into one for rooms and one for the rest.
+    # Note that the root element is included, which will link to the main page.
+    sitemaps = {
+        "room": [],
+        "other": [],
+    }
+
+    # Look whether there are currently online sitemaps for the provided
+    # sitemaps name. In case there aren't, we assume this sitemap is new,
+    # and all entries will be marked as changed.
+    old_sitemaps = _download_online_sitemaps(sitemaps.keys())
+
+    changed_count = 0
+    for _id, entry in new_data.items():
+        sitemap_name = entry["type"] if entry["type"] in sitemaps else "other"
+
+        url = f"https://nav.tum.sexy/view/{_id}"
+
+        if _id not in old_data or entry != old_data[_id]:
+            last_changed = datetime.now()
+            changed_count += 1
+        else:
+            # Try to look up the last changed date in the old sitemap
+            last_changed = old_sitemaps.get(sitemap_name, {})\
+                                       .get(url, datetime.now())
```

Comment:
The only case when `old_sitemaps[sitemap_name]` does not exist is, when the sitemap is new, and thus there is not old sitemap for that name. That applies to this PR for example, because currently there is no sitemap online.

The second point is correct though, it should increment the counter.